internal/controller/etcdopstask/reconcile_delete.go

```diff
@@ -0,0 +1,146 @@
+
+	// If task is completed but not marked for deletion, wait for TTL expiry
+	if task.IsCompleted() {
+
+		if !task.HasTTLExpired() {
+			logger.Info("Task completed but TTL not expired yet, will requeue after TTL", "ttlSeconds", task.Spec.TTLSecondsAfterFinished)
+			return ctrlutils.ReconcileAfter(task.GetTimeToExpiry(), "Task completed, waiting for TTL to expire")
+		}
+		logger.Info("Task TTL expired, proceeding with deletion")
+	}
+
+	deletionStepFns := []StepFunction{
+		{
+			StepName: "CleanupTaskResources",
+			StepFunc: r.cleanupTaskResources,
+		},
+		{
+			StepName: "RemoveTaskFinalizer",
+			StepFunc: r.removeTaskFinalizer,
+		},
+		{
+			StepName: "RemoveTask",
+			StepFunc: r.removeTask,
+		},
+	}
+	for _, step := range deletionStepFns {
+		logger.Info("Executing deletion step", "stepName", step.StepName)
+		result := step.StepFunc(ctx, client.ObjectKeyFromObject(task), taskHandler)
+		if ctrlutils.ShortCircuitReconcileFlow(result) {
+			logger.Info("Short-circuiting deletion flow", "stepName", step.StepName, "result", result)
+			return result
+		}
+	}
+	logger.Info("Deletion flow completed for EtcdOpsTask")
+	return ctrlutils.DoNotRequeue()
+}
+
+// cleanupTaskResources calls the Cleanup method of the task handler.
+func (r *Reconciler) cleanupTaskResources(ctx context.Context, taskObjKey client.ObjectKey, taskHandler opsTask.Handler) ctrlutils.ReconcileStepResult {
+	logger := r.logger.WithValues("namespace", taskObjKey.Namespace, "name", taskObjKey.Name)
+	logger.Info("Cleaning up task resources")
+
+	task, err := r.getTask(ctx, taskObjKey)
+	if err != nil {
+		return ctrlutils.ReconcileWithError(err)
+	}
+
+	if task.Status.State != nil && *task.Status.State == v1alpha1.TaskStateRejected {
+		logger.Info("Task is rejected, skipping cleanup")
+		// no-op cleanup
+		// Cases where task is rejected:
+		// 1. Task is not supported by the controller
+		// 2. Task admit failed
+		// In these cases, we don't want to call the cleanup method of the task handler. Since there is no cleanup to be done, we can skip this step.
+		if err := r.recordLastOperation(ctx, taskObjKey, v1alpha1.OperationPhaseCleanup, v1alpha1.OperationStateCompleted, ""); err != nil {
+			return ctrlutils.ReconcileWithError(err)
+		}
+		return ctrlutils.ContinueReconcile()
+	}
+
+	if err := r.recordLastOperation(ctx, taskObjKey, v1alpha1.OperationPhaseCleanup, v1alpha1.OperationStateInProgress, ""); err != nil {
+		return ctrlutils.ReconcileWithError(err)
+	}
+
+	result := taskHandler.Cleanup(ctx)
+	if result == nil {
+		err := druiderr.WrapError(
+			fmt.Errorf("cleanup returned nil TaskResult; this is a bug in the TaskHandler implementation"),
+			ErrNilResult,
+			string(v1alpha1.OperationPhaseCleanup),
+			"cleanup returned nil TaskResult; this is a bug in the TaskHandler implementation",
+		)
+		_ = r.recordLastError(ctx, taskObjKey, err)
+		return ctrlutils.ReconcileWithError(err)
+	}
+
+	if result.Completed {
```

Comment:
I am a bit confused here - when doing a cleanup if it is marked as `completed` then why do you check for `error` because in the else block you already check if the result was in error? The task can either be completed successfully or be in error. If there is an error it is not complete. If this is true, then having `Completed` and `Error` in `Result` is not really required, right?